modules/programs/fd.nix

```
@@ -0,0 +1,77 @@
+{ config, lib, pkgs, ... }:
+with lib; {
+  meta.maintainers = [ maintainers.uncenter ];
+
+  options.programs.fd = {
+    enable = mkEnableOption
+      "fd, a simple, fast and user-friendly alternative to {command}`find`";
+
+    enableBashIntegration = mkEnableOption "Bash integration" // {
+      default = true;
+    };
+
+    enableZshIntegration = mkEnableOption "Zsh integration" // {
+      default = true;
+    };
+
+    enableFishIntegration = mkEnableOption "Fish integration" // {
+      default = true;
+    };
+
+    enableIonIntegration = mkEnableOption "Ion integration" // {
+      default = true;
+    };
+
+    enableNushellIntegration = mkEnableOption "Nushell integration";
+
+    ignores = mkOption {
+      type = types.listOf types.str;
+      default = [ ];
+      example = [ ".git/" "*.bak" ];
+      description = "List of paths that should be globally ignored.";
+    };
+
+    hidden = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        Search hidden files and directories ({option}`--hidden` argument).
+      '';
+    };
+
+    extraOptions = mkOption {
+      type = types.listOf types.str;
+      default = [ ];
+      example = [ "--no-ignore" "--absolute-path" ];
+      description = ''
+        Extra command line options passed to fd.
+      '';
+    };
+
+    package = mkPackageOption pkgs "fd" { };
+  };
+
+  config = let
+    cfg = config.programs.fd;
+
+    args = escapeShellArgs (optional cfg.hidden "--hidden" ++ cfg.extraOptions);
+
+    optionsAlias = { fd = "fd ${args}"; };
+  in mkIf cfg.enable {
+    home.packages = [ cfg.package ];
+
+    programs.bash.shellAliases = optionsAlias;
+
+    programs.zsh.shellAliases = optionsAlias;
+
+    programs.fish.shellAliases = optionsAlias;
+
+    programs.ion.shellAliases = optionsAlias;
+
+    programs.nushell.shellAliases = optionsAlias;
+
+    xdg.configFile."fd/ignore" = mkIf (cfg.ignores != [ ]) {
+      text = concatStringsSep "\n" cfg.ignores + "\n";
+    };
+  };
```

```suggestion
  config = let
    cfg = config.programs.fd;

    args = escapeShellArgs (optional cfg.hidden "--hidden" ++ cfg.extraOptions);

    ignores = mkIf (cfg.ignores != [ ]) {
      text = concatStringsSep "\n" cfg.ignores + "\n";
    };

    optionsAlias = { fd = "fd ${args}"; };
  in mkIf cfg.enable (mkMerge [
    {
      home.packages = [ cfg.package ];

      programs.bash.shellAliases = optionsAlias;

      programs.zsh.shellAliases = optionsAlias;

      programs.fish.shellAliases = optionsAlias;

      programs.ion.shellAliases = optionsAlias;

      programs.nushell.shellAliases = optionsAlias;
    }
    (mkIf (!config.home.preferXdgDirectories) {
      home.file.".fdignore".text = ignores;
    })
    (mkIf (!config.home.preferXdgDirectories) {
      xdg.configFile."fd/ignore" = ignores;
    })
  ]);
```

---

We could add something like this to choose between `~/.fdignore`  and `$XDG_CONFIG_HOME/fd/ignore` using `config.home.preferXdgDirectories`.

---

From `fd`'s man page:

> ### `XDG_CONFIG_HOME`, `HOME`
> Used  to  locate the global ignore file. If `XDG_CONFIG_HOME` is set, use `$XDG_CONFIG_HOME/fd/ignore`.  Otherwise, use `$HOME/.config/fd/ignore`.

I'm fine with that I guess but we should write to `home.file.".config/fd/ignore"` instead?

---

I guess you could have
``` nix
let configPath = if config.home.preferXdgDirectories then "${config.xdg.configHome}/fd/ignore" else ".fdignore";
```

and then do

``` nix
home.file.${configPath}.text = ignores;
```

---

```suggestion
  config = let
    cfg = config.programs.fd;

    args = escapeShellArgs (optional cfg.hidden "--hidden" ++ cfg.extraOptions);

    optionsAlias = { fd = "fd ${args}"; };

    ignores = mkIf (cfg.ignores != [ ]) {
      text = concatStringsSep "\n" cfg.ignores + "\n";
    };

    configPath = if config.home.preferXdgDirectories then
      "${config.xdg.configHome}/fd/ignore"
    else
      ".fdignore";
  in mkIf cfg.enable {
    home.packages = [ cfg.package ];

    programs.bash.shellAliases = optionsAlias;

    programs.zsh.shellAliases = optionsAlias;

    programs.fish.shellAliases = optionsAlias;

    programs.ion.shellAliases = optionsAlias;

    programs.nushell.shellAliases = optionsAlias;

    home.file.${configPath}.text = ignores;
  };
}
```

This indeed looks cleaner.

But if `fd` requires the `XDG_CONFIG_HOME` variable to look for `~/.config/fd/ignores` I guess we should also check for `config.xdg.enable`? Would not it be better to always enable `config.xdg.enable` when setting `home.preferXdgDirectories = true`?

---

> Used to locate the global ignore file. If XDG_CONFIG_HOME is set, use $XDG_CONFIG_HOME/fd/ignore. Otherwise, use $HOME/.config/fd/ignore.

That's just following the XDG spec no? So always storing it using `xdg.configFile` should just work.

IMO, `preferXdgDirectories` should be avoided in most cases, and used for programs that have no understanding of XDG at all and need e.g: an alias to change the config location.

Anyways, I don't even see `~/.fdignore` mentionned in the manual, just `~/.config/fd/ignore` (the default `XDG_CONFIG_HOME` location).

---

> Anyways, I don't even see `~/.fdignore` mentionned in the manual, just `~/.config/fd/ignore` (the default `XDG_CONFIG_HOME` location).

That's what I've been trying to say! 😅

---

The [documentation](https://github.com/sharkdp/fd#excluding-specific-files-or-directories) only mentions the xdg config home location. So I think this comment can be safely ignored.

---

Oh yes indeed! I misunderstood your comment @uncenter, ignore my comments.